File: src/main/java/ai/aimachineserver/domain/games/tictactoe/Board.java
package ai.aimachineserver.domain.games.tictactoe;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Board {
    public final static int DEFAULT_SIZE = 3;
    public final static int EMPTY_FIELD_VALUE = 0;
    public final int size;
    private final int[][] allFieldValues;

    public Board(int size) {
        this.size = size;
        allFieldValues = new int[size][size];
        setAllFieldsEmpty();
    }

    public Board() {
        this.size = DEFAULT_SIZE;
        allFieldValues = new int[size][size];
        setAllFieldsEmpty();
    }

    int[][] getAllFieldValues() {
        return allFieldValues;
    }

    void setFieldValue(int rowIndex, int colIndex, int fieldValue) {
        allFieldValues[rowIndex][colIndex] = fieldValue;
    }

    List<BoardCoords> getAvailableFieldIndices() {
        List<BoardCoords> availableFieldIndices = new ArrayList<>();
        for (int rowIndex = 0; rowIndex < allFieldValues.length; rowIndex++) {
            for (int colIndex = 0; colIndex < allFieldValues[rowIndex].length; colIndex++) {
                if (allFieldValues[rowIndex][colIndex] == EMPTY_FIELD_VALUE) {
                    availableFieldIndices.add(new BoardCoords(rowIndex, colIndex));
                }
            }
        }
        return availableFieldIndices;
    }

    boolean isFieldAvailable(int rowIndex, int colIndex) {
        return allFieldValues[rowIndex][colIndex] == EMPTY_FIELD_VALUE;
    }

    void clearAllFields() {
        setAllFieldsEmpty();
    }

    private void setAllFieldsEmpty() {
        for (int[] row : allFieldValues) {
            Arrays.fill(row, EMPTY_FIELD_VALUE);
        }
    }
}

File: src/main/kotlin/ai/aimachineserver/domain/games/tictactoe/Board.kt
package ai.aimachineserver.domain.games.tictactoe

class Board(val size: Int = DEFAULT_SIZE) {

    companion object {
        const val DEFAULT_SIZE = 3
        const val EMPTY_FIELD_VALUE = 0
    }

    private val allFieldValues: Array<IntArray> = Array(size) { IntArray(size) { EMPTY_FIELD_VALUE } }

    fun getAllFieldValues() = allFieldValues

    fun setFieldValue(rowIndex: Int, colIndex: Int, fieldValue: Int) {
        allFieldValues[rowIndex][colIndex] = fieldValue
    }

    fun getAvailableFieldIndices(): List<BoardCoords> {
        val availableFieldIndices = mutableListOf<Pair<Int, Int>>()
        allFieldValues.forEachIndexed { rowIndex, row ->
            row.forEachIndexed { colIndex, fieldValue ->
                if (fieldValue == EMPTY_FIELD_VALUE) {
                    availableFieldIndices.add(Pair(rowIndex, colIndex))
                }
            }
        }
        return availableFieldIndices.map { BoardCoords(it.first, it.second) }
    }

    fun isFieldAvailable(rowIndex: Int, colIndex: Int) = allFieldValues[rowIndex][colIndex] == EMPTY_FIELD_VALUE

    fun clearAllFields() = allFieldValues.forEachIndexed { rowIndex, row ->
        row.indices.forEach { colIndex ->
            allFieldValues[rowIndex][colIndex] = EMPTY_FIELD_VALUE
        }
    }
}
